Add copy button

cfurrow opened the pull request:
Add a "copy" button that can copy the current code block into the clipboard, per this feature request https://github.com/yusanshi/embed-like-gist/issues/9

- Adds clipboard.js to the 404.html page
  - I'm not opposed to using something like the 404.html page uses versus embedding a new script. 
- Adds "Show Copy Button" checkbox to the main form
- When "show copy button" is on, a "copy" button appears in the embedded code block.
  - Clicking the button copies the contents of the code block to the clipboard (via clipboard.js)

Dark theme:
![image](https://user-images.githubusercontent.com/149965/113512687-c53e5780-9533-11eb-986f-64b9f385b26b.png)

Light them:
![image](https://user-images.githubusercontent.com/149965/113512762-164e4b80-9534-11eb-8749-5b0287167db0.png)



### TODO:
- Style button (or make a link)
- Use a "copy" icon for the button?

cfurrow committed:
Look for showCopy and showShare options, and render buttons if present

cfurrow committed:
Include clipboard.js, setup on copy button, remove share button (for now)

cfurrow committed:
Update clipboardjs to latest

cfurrow committed:
Add showCopy setting to localStorage

cfurrow committed:
Add very basic styles to copy button and toolbar area

cfurrow committed:
Remove TODO comment

cfurrow committed:
Fix whitespace

yusanshi commented:
Hi @cfurrow ,

Thanks for your work!

I tested it and had found an issue, that is, when the code is embedded on third-party pages, then `clipboard.js` is in fact not imported (I just added the [Test part](https://github.com/yusanshi/embed-like-gist#test) to `README.md` and you can try the 4th of that part).

To make it work, it seems that we should do something like <https://github.com/yusanshi/embed-like-gist/blob/master/embed.js#L76-L87>.

However, to me, on our own pages (i.e., `404.html`), we can import any libraries we want but on others' pages (i.e., the pages where the code is embedded), I prefer to import the least libraries for two reasons:
1. speed up the loading.
2. influence (or pollute, honestly) the target page less.

I believe the `highlight.js` and `highlightjs-line-numbers.js` are required but not `clipboard.js`.

To copy the text we can write it ourselves (https://github.com/yusanshi/embed-like-gist/blob/master/404.html#L294-L323). To show a popover on copying I believe implementing it with pure CSS is also feasible.

cfurrow commented:
Hello @yusanshi, you're absolutely right! I forgot to test this on an actual page! I wanted to do a little more cleanup on this anyway (that's why it's still left in draft mode), but I will definitely use the copy code you already had within the test page, and embed it in the JS.

Thanks!

cfurrow committed:
Remove clipboard.js from 404

cfurrow committed:
Remove clipboard.js code from embed.js

cfurrow committed:
Pull in copyTextToClipboard and fallbackCopyTextToClipboard from 404 to embed.js

cfurrow committed:
Add click even handler to copyButton, grab innerText from 'emgithub-code'

cfurrow committed:
Set an :active style (green bg) for copy button

cfurrow committed:
Use the codeText variable when copying to clipboard

We already have the text content, why not use that instead of re-scraping the HTML for it?

cfurrow commented:
I'm running into an odd issue on my "test" page where I have three embed.js examples:

```html
<div>
  <h1>Test 1 (embed.js)</h1>
  <script src="/embed.js?target=https%3A%2F%2Fgithub.com%2Fyusanshi%2Fembed-like-gist%2Fblob%2F6c0ad6e401c7af9b5d6aff29af18f71b3105d803%2Fembed.js%23L1-L10&style=github&showBorder=on&showLineNumbers=on&showFileMeta=on&showCopy=on"></script>
</div>

<div>
  <h1>Test 2 (highlightjs badge)</h1>
  <script src="/embed.js?target=https%3A%2F%2Fgithub.com%2FRickStrahl%2Fhighlightjs-badge%2Fblob%2Fmaster%2Fhighlightjs-badge.js%23L50-L94&style=github&showBorder=on&showLineNumbers=on&showFileMeta=on&showCopy=on"></script>
</div>

<div>
  <h1>Test 3 (wave.js)</h1>
  <script src="/embed.js?target=https%3A%2F%2Fgithub.com%2Fcfurrow%2Fphaser-wave-example%2Fblob%2Fmaster%2Fwave.js&style=github&showBorder=on&showLineNumbers=on&showFileMeta=on&showCopy=on"></script>
</div>
```

When everything is fully rendered, the leading whitespace seems to be removed completely, and I've tracked it down to happening somewhere inside of highlightjs-line-numbers.min.js.

Have you seen this before? It doesn't seem to happen on the test page. I tried a few times with just one embed.js example on my test page, and it still removes the leading whitespace when line numbers are enabled.

## With Line Numbers ❌
![image](https://user-images.githubusercontent.com/149965/114039321-a3edac00-9850-11eb-975a-dbbad0378782.png)


## Without Line Numbers ✅
![image](https://user-images.githubusercontent.com/149965/114039345-a8b26000-9850-11eb-9d23-91dbf0a62ff9.png)

yusanshi commented:
I use this example and it works well:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Test</title>
  </head>
  <body style="padding:5rem">
    <div>
      <h1>Test 1 (With line numbers)</h1>
      <script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Fyusanshi%2Fembed-like-gist%2Fblob%2F6c0ad6e401c7af9b5d6aff29af18f71b3105d803%2Fembed.js%23L1-L10&style=github&showBorder=on&showLineNumbers=on&showFileMeta=on&showCopy=on"></script>
    </div>
    <div>
      <h1>Test 2 (Without line numbers)</h1>
      <script src="https://emgithub.com/embed.js?target=https%3A%2F%2Fgithub.com%2Fyusanshi%2Fembed-like-gist%2Fblob%2F6c0ad6e401c7af9b5d6aff29af18f71b3105d803%2Fembed.js%23L1-L10&style=github&showBorder=on&showFileMeta=on&showCopy=on"></script>
    </div>
  </body>
</html>
```

**Chrome v89 on Ubuntu**

![image](https://user-images.githubusercontent.com/36265606/114166508-64ae8080-9960-11eb-9316-ab1eabda8245.png)


**Firefox v87 on Ubuntu**

![image](https://user-images.githubusercontent.com/36265606/114166605-8a3b8a00-9960-11eb-8fa3-316c1df55a1a.png)


What's the OS and browser you're using?

And could you inspect the page and check whether there is a `white-space: pre` style?
![image](https://user-images.githubusercontent.com/36265606/114167023-046c0e80-9961-11eb-8fd2-332f15629f06.png)


<https://developer.mozilla.org/en-US/docs/Web/CSS/white-space>

cfurrow committed:
Set code.textContent once, and re-use codeText var to hold spliced content

This allows us to use the "rendered" code as the source for copying. The closure should keep this codeText value around when the user decides to click the "copy" button for the code block.

cfurrow commented:
Thanks for that info. Your test file seems to work for me as well. I am not sure what may have been wrong with my example, as it was a bare-bones HTML page on the Brave Browser (Chromium) on macOS. I even saw the same issue with Safari.

Now that it seems to work and the whitespace correct, I'm closer to wrapping up this PR and getting it reviewed. 

![CleanShot 2021-04-09 at 11 25 57](https://user-images.githubusercontent.com/149965/114203867-8ee34c80-9926-11eb-92bd-1e775e29a832.gif)

cfurrow commented on embed.js:
```diff
@@ -136,10 +185,9 @@ function embedCodeToTarget(targetDiv, codeText, showBorder, showLineNumbers, sho
   code.classList.add(lang);
   if (startLine > 0) {
     codeTextSplit = codeText.split("\n");
-    code.textContent = codeTextSplit.slice(startLine - 1, endLine).join("\n");
-  } else {
-    code.textContent = codeText;
+    codeText = codeTextSplit.slice(startLine - 1, endLine).join("\n");
   }
+  code.textContent = codeText;
```
Using the codeText variable for the "copy" action, and the closure will keep the variable in the correct scope in the case of multiple embed.js examples on a single page.

cfurrow committed:
Remove unused targetDivID

cfurrow committed:
Remove unused emgithub-code CSS class

cfurrow commented:
Thanks again for all the time you've provided in helping me work through this issue, @yusanshi. I'm wondering if you could elaborate on your prior comment:

> To show a popover on copying I believe implementing it with pure CSS is also feasible.

Are you referring to potentially rendering the "copy" button as a [pseudo-element](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements)? Or did you have something else in mind versus what I've done, which is a lot of JS DOM manipulation/element creation.

yusanshi committed:
Adjust CSS attributes

yusanshi committed:
Rename for consistency

yusanshi committed:
Merge branch 'master' into copy-share-buttons

yusanshi committed:
Add the loading of 'showCopy'

yusanshi committed:
Trim whitespace

yusanshi commented:
Hi @cfurrow ,
After googling I realized that popover with pseudo-element, pseudo-class and some JavaScript isn't a good idea. A simple implementation doesn't look well, and a beautiful one may need a lot of work considering the compatibility on different OS and browsers.

Well, GitHub doesn't show the popover, too. I think currently our background color changing on `:active` is enouth 😆😆
![image](https://user-images.githubusercontent.com/36265606/114515186-aa27c200-9c6e-11eb-98b9-96048dc0aae2.png)

In my commits just now, I mainly adjusted some CSS attributes, including font, padding, etc. You can check it and after that, I think it's ready for mering.

Thanks for your work!

cfurrow commented:
Those changes look good to me! Thanks for taking the time to tweak it.

Looks good to merge.
